packages/compile-vyper/index.js

```diff
@@ -178,53 +171,101 @@ async function compileAllNoJson({ sources, options, version }) {
 
 const Compile = {
   // Check that vyper is available then forward to internal compile function
-  async sources({ sources = [], options }) {
+  async sources({ sources = {}, options }) {
     options = Config.default().merge(options);
-    // filter out non-vyper paths
-    const vyperFiles = sources.filter(
-      path => minimatch(path, VYPER_PATTERN, { dot: true })
-    );
-    const vyperFilesStrict = vyperFiles.filter(
-      path => minimatch(path, VYPER_PATTERN_STRICT, { dot: true })
+    const paths = Object.keys(sources);
+    const vyperFiles = paths.filter(path =>
+      minimatch(path, VYPER_PATTERN_STRICT, { dot: true })
     );
 
     // no vyper files found, no need to check vyper
     // (note that JSON-only will not activate vyper)
-    if (vyperFilesStrict.length === 0) {
+    if (vyperFiles.length === 0) {
       return { compilations: [] };
     }
 
-    const { version, json } = await checkVyper();
-    return await compileAll({
-      sources: vyperFiles,
-      options,
-      version,
-      useJson: json
-    });
+    Compile.display(vyperFiles, options);
+    const { version, json: useJson } = await checkVyper();
+    if (useJson) {
+      return compileJson({ sources, options, version });
+    } else {
+      return await compileNoJson({ paths: vyperFiles, options, version });
+    }
   },
 
-  //since we don't have an imports analyzer for Vyper
-  //yet, we'll just treat this the same as all; this will
-  //need to be revisited once we have an import parser for Vyper
-  async sourcesWithDependencies({ paths: _paths = [], options }) {
-    return await Compile.all({ options });
+  async sourcesWithDependencies({ paths = [], options }) {
+    options = Config.default().merge(options);
+    debug("paths: %O", paths);
+    const vyperFilesStrict = paths.filter(path =>
+      minimatch(path, VYPER_PATTERN_STRICT, { dot: true })
+    );
+    debug("vyperFilesStrict: %O", vyperFilesStrict);
+
+    // no vyper targets found, no need to check Vyper
+    if (vyperFilesStrict.length === 0) {
+      return { compilations: [] };
+    }
+
+    const { allSources, compilationTargets } = await requiredSources(
+      options.with({
+        paths: vyperFilesStrict,
+        base_path: options.contracts_directory
+      })
+    );
+
+    debug("allSources: %O", allSources);
+    debug("compilationTargets: %O", compilationTargets);
+    const vyperTargets = compilationTargets.filter(path =>
+      minimatch(path, VYPER_PATTERN_STRICT, { dot: true })
+    );
+
+    // no vyper targets found, no need to activate Vyper
+    if (vyperTargets.length === 0) {
+      return { compilations: [] };
+    }
+
+    //having gotten the sources from the resolver, we invoke compileJson
+    //ourselves, rather than going through Compile.sources()
+    Compile.display(compilationTargets, options);
+
+    const { version, json: useJson } = await checkVyper();
+
+    if (useJson) {
+      return compileJson({
```

Comment:
What do you think about these two methods `compileJson` and `compileNoJson` having a different signature?

Reply:
It's a bit awkward, to be sure, but I'm not sure how to do better here.  In the no-JSON case, we're basically stuck passing paths to the compiler rather than full source texts.  So, I think it makes sense in context that they have different signatures?  It's not like either will be called directly by the outside.

Reply:
Actually, somewhat tangential thought -- I'm wondering if `sources` should just throw if JSON isn't available.  Because `sources` isn't called by `WorkflowCompile`, and, like, it's not going to do what the person using it likely expects.  I'm thinking maybe it should just throw... thoughts, @gnidan?